### PR TITLE
Milestone 5 - Friendship Functionalities and Testing - v1

### PR DESCRIPTION
**Pull request by ldelbel**
Hello TSE,

In this milestone I added the friendship functionality to the social media scaffold, with the following components:

- Friendship model and Users/Users many-to-many relations
- Friendships and Friends controllers and views
- Button update automatization for send, cancel, accept, and decline friendship request, as well as friend/unfriend

The following tests were made: 
- Request tests using RSpec
- Integration tests using RSpec/Capybara
- For most  tests, I used Database-Cleaner

I kindly request the TSE to  review my code

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @ldelbel :clap: However, some changes need to be applied in order to get approval for this milestone. Check :arrow_down: 

#### General (required) :warning: 
* [x] use rubocop.yml version provided by microverse, don't apply your own changes.

#### Behaviour (required) :warning: 
* [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
#### Implementation (required) :warning: 
- Below, I suggest a more robust, clean and readable way to write your associations.
* [x] User model `has_many :friendships` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] User model `has_many :inverted_friendships, class_name: "Friendship", foreign_key: "friend_id"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Friendship model `belongs_to :user` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Friendship model `belongs_to :friend, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -20,7 +20,8 @@ def create
   private
 
   def timeline_posts
-    @timeline_posts ||= Post.all.ordered_by_most_recent.includes(:user)
+    friendlist = 'user_id IN (?) OR user_id IN (?)', current_user.friends.pluck(:id), [current_user.id]
+    @timeline_posts ||= Post.where(friendlist).ordered_by_most_recent.includes(:user)
```
Comment (tirthajyoti-ghosh) [inline app/controllers/posts_controller.rb]: Use built-in Rails methods to prepare SQL queries. It's a really bad idea and a bad practice to build SQL queries yourself.

You can learn more about active record queries from [here](https://guides.rubyonrails.org/active_record_querying.html)

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

### However, we need to work on some changes before we can merge. Please check the comments below.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by rOluochKe — APPROVED** [review]
Hello @ldelbel 
### Well done, you got it right :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

**Pull request merged**